Refine ESQL docs handling of applies_to

craigtaverner opened the pull request:
This primarily splits the old `preview:true` warning from the newer `applies_to` approach. Since all of our current `applies_to` examples are actually just behaviour modifications of current functions, we do not use the official docs `{applies_to}` syntax. However there is code to make use of that in the case where we have an entirely new function which will appear in a new version.

## Examples:

### An example of a function which has both preview and applies to is:

<img width="921" alt="Screenshot 2025-03-28 at 12 51 19" src="https://github.com/user-attachments/assets/3eb0aeca-691d-4319-b4c0-439d55e3f094" />

There are many pre-existing examples of preview only, which we've left as is.

### An example of applies_to without preview is:

<img width="917" alt="Screenshot 2025-03-28 at 12 53 40" src="https://github.com/user-attachments/assets/7d17ce9c-7a76-4282-9c58-7b221bdf4628" />

craigtaverner committed:
Remove all warnings from DocsV3Support

craigtaverner committed:
Split applies_to from preview warning

craigtaverner committed:
Update examples test to use real examples rendering

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Values.java:
```diff
@@ -86,8 +84,7 @@ public class Values extends AggregateFunction implements ToAggregator {
             ::::""",
         type = FunctionType.AGGREGATE,
-        examples = @Example(file = "string", tag = "values-grouped"),
-        appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.PREVIEW) }
```
This file has both the old and new approaches to tech-preview, so I removed the new one until we make a final decision on how to deal with tech-preview warnings in general. They are all over the place, not just here, so this function was just an experiment to see what it looked like.

craigtaverner committed:
Merge branch 'main' into esql_docs_applies_to

alex-spies commented on docs/reference/query-languages/esql/_snippets/functions/layout/match.md:
```diff
@@ -2,12 +2,13 @@
 are not subject to the support SLA of official GA features.
+:::
 
+:::{note}
+###### COMING 9.1.0
```
I just want to say that this looks less confusing to me than having the `COMING 9.1.0` inside the tech preview warning.

leemthompo replied:
💯

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java:
```diff
@@ -213,19 +230,53 @@ private static Map.Entry<String, OperatorConfig> operatorEntry(
+    private TempFileWriter tempFileWriter;
 
     private DocsV3Support(String category, String name, Class<?> testClass, Supplier<Map<List<DataType>, DataType>> signatures) {
+        this(category, name, definition(name), testClass, signatures);
```
nit: duplicated logic could be avoided.
```suggestion
        this(category, name, null, testClass, signatures);
```

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java:
```diff
@@ -381,20 +425,37 @@ private FunctionDocsSupport(String name, Class<?> testClass) {
             super("functions", name, testClass, () -> AbstractFunctionTestCase.signatures(testClass));
```
nit: this could use the new, more general c'tor below.

craigtaverner replied:
I'd rather not add null parameters in more places. I committed your other suggestion because the null parameter was resolved in code just a few lines below, but in this case the null is processed further away, so I'd rather not have it visible here at all.

craigtaverner committed:
Fix EsqlNodeSubclassTests failures

Any class that extends Function needs to correctly implement replaceChildren and info() methods, because it will be found and tested!

alex-spies reviewed (approved):
Heya, I didn't deeply review the java code but the generated output looks good to me (as a non-docs person).

craigtaverner committed:
Merge branch 'esql_docs_applies_to' of github.com:craigtaverner/elasticsearch into esql_docs_applies_to

craigtaverner committed:
Update x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

craigtaverner committed:
Merge branch 'main' into esql_docs_applies_to

craigtaverner committed:
Refined applies_to, adding serverless and removing some version info

craigtaverner committed:
Merge branch 'esql_docs_applies_to' of github.com:craigtaverner/elasticsearch into esql_docs_applies_to

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_docs_applies_to

craigtaverner committed:
Remove applies_to from LOOKUP JOIN and MV_EXPAND

alex-spies commented on docs/reference/query-languages/esql/esql-commands.md:
````diff
@@ -666,10 +666,6 @@ FROM employees
 
 ::::{warning}
-```{applies_to}
-stack: preview 9.0, coming 9.1
-serverless: preview
````
I agree that this maybe shouldn't be in the tech preview warning, but I don't know if these tags should be removed altogether. Maybe @leemthompo can suggest a better position?

craigtaverner commented on docs/reference/query-languages/esql/esql-commands.md:
````diff
@@ -753,10 +749,6 @@ FROM Left
 
 ::::{warning}
-```{applies_to}
-stack: preview 9.0, coming 9.1
````
I added these originally as examples to discuss with people, but failed to remove them. Based on current discussions, we should remove any version statements, even if the feature is available in serverless, as there is no guarrantee if it will come out in any particular stateful version.

craigtaverner committed:
Use Elastic Stack instead of Stateful, and remove version from QSTR

craigtaverner committed:
Merge branch 'main' into esql_docs_applies_to

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  9.0  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 125835`

craigtaverner commented:
Manual backport in https://github.com/elastic/elasticsearch/pull/125923
